### PR TITLE
internal/server: refactor install into one package

### PR DESCRIPTION
**Pull request by krantzinator**
Foundation to addressing #672 and #612
There's a lot of diff from moving existing code around, so I'll highlight actual changes here:
- Moved the platforms to implement an `Installer` interface, replacing the `case` setup in `InstallCommand` to set the platform
- Took `serverinstall.Config` out of `InstallCommand` and gave each `Platform` its own config
  - This was the main place I went in circles with; I wanted each Platform to have its own config, and only care about the flags it needed to. I tried out a separate `serverinstall/config` package; toyed with using `serverconfig`; moved the `InstallFlags` pieces from `InstallCommand` to `serverinstall` to separate packages and back again. I also worked on this piece with a couple teammates, experimenting with different ways to have generic flags and install them to the scoped platforms as part of init. Ultimately I decided to go with the Platform-specific flags you see here in order to overcome the various hurdles discovered.

Small pieces:
- fixed a bug in the k8s statefulset; we weren't actually using the value set in the `pull-policy` flag, and updated the flag to have an updated value of `IfNotPresent` to match k8s default (`ImagePullPolicy: apiv1.PullPolicy(c.ImagePullPolicy),`
- threw a `strings.ToLower` on the platform name for UX

Copying Evan's example of including examples in lieu of a CLI testing framework:
```
$ waypoint-dev install -platform=nomad -nomad-dc=dc1 -accept-tos
✓ Nomad allocation ready
✓ Configuring server...
Waypoint server successfully installed and configured!
...

$ waypoint-dev install -platform=docker -accept-tos
✓ Installing Waypoint server to docker
✓ Server container started!
✓ Configuring server...
Waypoint server successfully installed and configured!
...

$ waypoint-dev install -platform=kubernetes -accept-tos
✓ Creating Kubernetes resources...
✓ Kubernetes StatefulSet reporting ready
✓ Waiting for Kubernetes service to become ready..
✓ Configuring server...
Waypoint server successfully installed and configured!
...
```

### CODE REVIEW

File: internal/serverinstall/docker.go [hunk 1]
```diff
@@ -14,14 +14,25 @@ import (
...
 	"github.com/hashicorp/waypoint/internal/serverconfig"
 )
 
-func InstallDocker(
-	ctx context.Context, ui terminal.UI, scfg *Config) (
+type DockerInstaller struct {
+	Config DockerConfig
+}
+
+type DockerConfig struct {
```
Comment (mitchellh) [inline internal/serverinstall/docker.go:30]: I think we can lowercase all these and keep them unexported (same with Nomad/K8S).

File: internal/serverinstall/serverinstall.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+	"context"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/waypoint-plugin-sdk/terminal"
+	"github.com/hashicorp/waypoint/internal/clicontext"
+	"github.com/hashicorp/waypoint/internal/pkg/flag"
+	pb "github.com/hashicorp/waypoint/internal/server/gen"
+)
+
+type Installer interface {
```
Comment (mitchellh) [inline internal/serverinstall/serverinstall.go:13]: Doc comment and also recommend commenting the interface methods. Take a look at some of our other interfaces we _usually_ do this.

File: internal/serverinstall/serverinstall.go [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+	Install(ctx context.Context, ui terminal.UI, log hclog.Logger) (*clicontext.Config, *pb.ServerConfig_AdvertiseAddr, string, error)
+	InstallFlags(*flag.Set)
+}
+
+var Platforms map[string]Installer = map[string]Installer{
```
Comment (mitchellh) [inline internal/serverinstall/serverinstall.go:18]: You should be able to just do this... but test it...

```suggestion
var Platforms = map[string]Installer{
```

**Review by mitchellh — CHANGES_REQUESTED** [review]
Yeahhhhhhh this is way cleaner.

Just a couple tiny comments but this is good work.

File: internal/serverinstall/serverinstall.go (hunk 2 above)
Comment (krantzinator) [reply internal/serverinstall/serverinstall.go:13]: Ah right, of course!

**Review by mitchellh — APPROVED** [review]

**Review by briancain — APPROVED** [review]
:package: :tada:

**Pull request merged**